quarkus/service/src/test/java/org/apache/polaris/service/quarkus/task/TaskRecoveryManagerTest.java

```diff
@@ -0,0 +1,213 @@
+              // no-op
+            }
+          };
+      TestServices testServices = TestServices.builder().realmContext(realmContext).build();
+      TaskExecutorImpl taskExecutor =
+          new TaskExecutorImpl(
+              Runnable::run,
+              metaStoreManagerFactory,
+              buildTaskFileIOSupplier(fileIO),
+              testServices.polarisEventListener()) {
+            @Override
+            public void addTaskHandlerContext(long taskEntityId, CallContext callContext) {
+              int attempts =
+                  retryCounter
+                      .computeIfAbsent(String.valueOf(taskEntityId), k -> new AtomicInteger(0))
+                      .incrementAndGet();
+              if (attempts == 1) {
+                // no-op for first attempt to mock failure
+              } else {
+                super.addTaskHandlerContext(taskEntityId, callContext);
+              }
+            }
+          };
+      taskExecutor.init();
+
+      TableCleanupTaskHandler tableCleanupTaskHandler =
+          new TableCleanupTaskHandler(
+              taskExecutor,
+              metaStoreManagerFactory,
+              buildTaskFileIOSupplier(new InMemoryFileIO())) {};
+
+      TableIdentifier tableIdentifier =
+          TableIdentifier.of(Namespace.of("db1", "schema1"), "table1");
+      long snapshotId = 100L;
+      ManifestFile manifestFile =
+          TaskTestUtils.manifestFile(
+              fileIO, "manifest1.avro", snapshotId, "dataFile1.parquet", "dataFile2.parquet");
+      TestSnapshot snapshot =
+          TaskTestUtils.newSnapshot(fileIO, "manifestList.avro", 1, snapshotId, 99L, manifestFile);
+      String metadataFile = "v1-49494949.metadata.json";
+      StatisticsFile statisticsFile =
+          TaskTestUtils.writeStatsFile(
+              snapshot.snapshotId(),
+              snapshot.sequenceNumber(),
+              "/metadata/" + UUID.randomUUID() + ".stats",
+              fileIO);
+      TaskTestUtils.writeTableMetadata(fileIO, metadataFile, List.of(statisticsFile), snapshot);
+
+      // Step 2: Execute the initial cleanup task, where two child cleanup tasks are generated and
+      // executed the first time
+      TaskEntity task =
+          new TaskEntity.Builder()
+              .setName("cleanup_" + tableIdentifier)
+              .withTaskType(AsyncTaskType.ENTITY_CLEANUP_SCHEDULER)
+              .withData(
+                  new IcebergTableLikeEntity.Builder(tableIdentifier, metadataFile)
+                      .setName("table1")
+                      .setCatalogId(1)
+                      .setCreateTimestamp(100)
+                      .build())
+              .build();
+      addTaskLocation(task);
+      Assertions.assertThatPredicate(tableCleanupTaskHandler::canHandleTask).accepts(task);
+      tableCleanupTaskHandler.handleTask(task, callCtx);
+
+      // Step 3: Verify that the generated child tasks were registered, ATTEMPT_COUNT = 2
+      timeSource.add(Duration.ofMinutes(10));
```

Comment:
I, personally, found this very hard to follow - even with the comments. I would highly recommend making the comments much more verbose here to allow the full flow of logic (what is happening with which task and why) to be communicated to a reader who may not be an expert at this particular type of task or tasks in general.